src/domain/accounts/counterfactual-safes/counterfactual-safes.repository.ts

```diff
@@ -0,0 +1,138 @@
+    chainId: string;
+    predictedAddress: `0x${string}`;
+  }): Promise<CounterfactualSafe> {
+    if (!args.authPayload.isForSigner(args.address)) {
+      throw new UnauthorizedException();
+    }
+    const [account] = await Promise.all([
+      this.accountsRepository.getAccount({
+        authPayload: args.authPayload,
+        address: args.address,
+      }),
+      this.checkCounterfactualSafesIsEnabled({
+        authPayload: args.authPayload,
+        address: args.address,
+      }),
+    ]);
+    return this.datasource.getCounterfactualSafe({
+      account,
+      chainId: args.chainId,
+      predictedAddress: args.predictedAddress,
+    });
+  }
+
+  /**
+   * Gets or creates a Counterfactual Safe.
+   * Checks that the account has the CounterfactualSafes data setting enabled.
+   *
+   * If the Counterfactual Safe exists, it returns it.
+   * If the Counterfactual Safe does not exist, it's created.
+   */
+  async getOrCreateCounterfactualSafe(args: {
+    authPayload: AuthPayload;
+    address: `0x${string}`;
+    createCounterfactualSafeDto: CreateCounterfactualSafeDto;
+  }): Promise<CounterfactualSafe> {
+    if (!args.authPayload.isForSigner(args.address)) {
+      throw new UnauthorizedException();
+    }
+    const [account] = await Promise.all([
+      this.accountsRepository.getAccount({
+        authPayload: args.authPayload,
+        address: args.address,
+      }),
+      this.checkCounterfactualSafesIsEnabled({
+        authPayload: args.authPayload,
+        address: args.address,
+      }),
+    ]);
+
+    try {
+      return await this.datasource.getCounterfactualSafe({
+        account,
+        chainId: args.createCounterfactualSafeDto.chainId,
+        predictedAddress: args.createCounterfactualSafeDto.predictedAddress,
+      });
+    } catch (error) {
+      if (error instanceof NotFoundException) {
+        return this.datasource.createCounterfactualSafe({
+          account,
+          createCounterfactualSafeDto: args.createCounterfactualSafeDto,
+        });
+      }
+      throw error;
+    }
+  }
+
+  private async checkCounterfactualSafesIsEnabled(args: {
+    authPayload: AuthPayload;
+    address: `0x${string}`;
+  }): Promise<void> {
+    const counterfactualSafeDataType =
+      await this.checkCounterfactualSafeDataTypeIsActive();
+    const accountDataSettings =
+      await this.accountsRepository.getAccountDataSettings({
+        authPayload: args.authPayload,
+        address: args.address,
+      });
+    const counterfactualSafeSetting = accountDataSettings.find(
+      (setting) =>
+        setting.account_data_type_id === counterfactualSafeDataType.id,
+    );
+    if (!counterfactualSafeSetting?.enabled) {
+      this.loggingService.warn({
+        message: `Account ${args.address} does not have CounterfactualSafes enabled`,
+      });
+      throw new GoneException();
+    }
+  }
+
+  private async checkCounterfactualSafeDataTypeIsActive(): Promise<AccountDataType> {
```

Comment:
The result of this function shouldn't change until a new migration is applied. `this.accountsRepository.getDataTypes()` already involves permanent caching. I think that should be enough,  but this function could be memoized eventually.